features/graph/api/src/main/java/org/opennms/netmgt/graph/api/generic/GenericGraph.java

```diff
@@ -45,12 +45,17 @@
 import org.opennms.netmgt.graph.api.focus.Focus;
 import org.opennms.netmgt.graph.api.info.GraphInfo;
 
+import com.google.common.base.Strings;
 import com.google.common.collect.Lists;
 
 import edu.uci.ics.jung.graph.DirectedSparseGraph;
 
-// TODO MVR enforce namespace
-public class GenericGraph extends GenericElement implements Graph<GenericVertex, GenericEdge> {
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * The graph itself is not immutable (Vertices and Edges can be added and removed) but it's properties are. 
```

Comment:
Is there value in having an immutable copy of the graph? If so we could add a `immutableCopy()` method to this class that creates a new instance of something that extends `DirectedSparseGraph` and adds the vertices and edges to it then returns that. The extended class can override the `add*()` methods with to throw unsupported exceptions.

Here is an example of copying the graph https://github.com/OpenNMS/alec/pull/7/files#diff-96fac5cebdf16f3307c72076ff349d96R82

Reply:
Each `Generic*` element should have an immutable counterpart.